### DNA extraction and sequencing data pre-processing

Genomic DNA extractions were performed on a KingFisher Flex magnetic particle processor robot (Thermo Scientific) using 
a NucleoMag® 96 Plant kit (Macherey-Nagel Gmbh & Co.). A volume of 150 μl was used for elution. DNA concentrations were 
measured on a Dropsense (TRINEAN NV) using a DropPlate 96-S. Based on these measurements, the DNA’s of the same 
phenotype (_EF_, _IF_, _LF_ and _NF_) were equimolarly pooled to create four DNA samples for HiSeq sequencing. 
<!-- add group-4 high coverage genome? -->
The bulks were sequenced on the Illumina HiSeq X platform resulting in a raw coverage, calculated as 
the total number of read bases divided by the size of the reference genome [approx. 488Mbp; @Parkin:2014], in the 
range of 111x-123x per bulk, i.e. approximately 10x per individual. We used the BWA-MEM [@Li:2013] and SAMtools 
[@Li:2009] toolchain to map the reads for each bulk against the _Brassica oleracea_ TO1000 reference genome v2.1 of 
EnsemblPlants release 39, which we had filtered so that we mapped against chromosomes only. We retained most of the 
estimated raw coverage in the assemblies, yielding an average mapped coverage in the range of 99x-108x. We then
used GATK HaplotypeCaller [@McKenna:2010; @Depristo:2011] for variant (i.e. SNP and indel) calling, yielding the 
results summarized in Table 1.

| Phenotype | Bulk size | Total read bases (bp) | Total reads | GC(%) | Q20(%) | Q30(%) | Coverage a, b | Variants   |
|-----------|-----------|-----------------------|-------------|-------|--------|--------|---------------|------------|
| _EF_      | 11        | 60,303,831,724        | 399,363,124 | 36.93 | 95.02  | 89.25  | 123, 108      | 40,224,519 |
| _IF_      | 8         | 56,804,209,216        | 376,186,816 | 36.92 | 94.84  | 88.94  | 116, 103      | 43,785,856 |
| _LF_      | 11        | 54,587,863,530        | 361,509,030 | 37.14 | 96.00  | 91.34  | 112, 100      | 42,852,937 |
| _NF_      | 9         | 54,296,890,456        | 359,582,056 | 37.13 | 96.84  | 92.81  | 111, 99       | 42,213,427 |

_Table 1. Summary results of the sequencing of bulks of early (_EF_), intermediate (_IF_), late (_LF_) and
"non" flowering (_NF_, actually not flowering at time of DNA extraction) phenotypes. Bulk size refers to the
number of individuals pooled for that phenotype. Coverage is given as a) total read bases divided by reference 
genome size; and b) average mapped coverage._

> #### Supplementary info
> **Reference data** - The processed reference genome (sans scaffolds; chromosomes only) is available from the main 
> data submission of this study, deposited at Zenodo as [10.5281/zenodo.3402201](http://doi.org/10.5281/zenodo.3402201).
> The processed FASTA file, and the annotation (in GFF3 format), are contained within the [reference.zip](reference.zip)
> archive, and should be stored locally as:
> - `${DATA}/reference/Brassica_oleracea.v2.1.dna.toplevel.chromosomes.fa` - FASTA
> - `${DATA}/reference/Brassica_oleracea.v2.1.39.chr.gff3.gz` - GFF3
>
> **Short read data mapping** - Reference mapping was automated with [bwa.sh](../script/bwa.sh). Input files are 
> deposited at the SRA as project [PRJNA564368](https://www.ncbi.nlm.nih.gov/bioproject/PRJNA564368),
> containing the BioSamples `group-1-EF`, `group-2-IF`, `group-3-LF`, and `group-5-NF`. These should be placed in a
> local folder structure as follows:
> - `${DATA}/BSA/${SAMPLE}/${SAMPLE}_R1.fastq.gz` - Paired end 1, FASTQ
> - `${DATA}/BSA/${SAMPLE}/${SAMPLE}_R2.fastq.gz` - Paired end 2, FASTQ
>
> **[Genome alignments and variant calls]** SNP calling was automated with [snp.sh](../script/snp.sh). The resulting 
> alignments (sorted BAM files) and variants (compressed VCF files) are  available from Zenodo under the following 
> locations:
> - EF [10.5281/zenodo.3401590](http://doi.org/10.5281/zenodo.3401590)
> - IF [10.5281/zenodo.3401703](http://doi.org/10.5281/zenodo.3401703)
> - LF [10.5281/zenodo.3401731](http://doi.org/10.5281/zenodo.3401731)
> - NF [10.5281/zenodo.3402013](http://doi.org/10.5281/zenodo.3402013)
> 
> In each case, the files in these Zenodo submissions should be organized locally as:
> - `${DATA}/BSA/${SAMPLE}/${SAMPLE}_pe.sorted.bam` - Genome alignment, BAM
> - `${DATA}/BSA/${SAMPLE}/${SAMPLE}_pe.sorted.bam.RG.vcf.gz` - Variant call file, VCF