### DNA extraction and sequencing data pre-processing

DNA extraction was done by ...
Library preparation was done using TruSeq DNA PCR Free library kit ...
Equimolar concentrations of the F1 genotypes were then pooled into the four bulks _EF_, _IF_, _LF_.
The bulks were sequenced on the Illumina HiSeq X platform resulting in a raw coverage, calculated as 
the total number of read bases divided by the size of the reference genome [approx. 488Mbp; @Parkin:2014], in the 
range of 111x-123x per bulk, i.e. approximately 10x per individual. We used the BWA-MEM [@Li:2013] and SAMtools 
[@Li:2009] toolchain to map the reads for each bulk against the _Brassica oleracea_ TO1000 reference genome v2.1 of 
EnsemblPlants release 39, which we had filtered so that we mapped against chromosomes only. We retained most of the 
estimated raw coverage in the assemblies, yielding an average mapped coverage in the range of 99x-108x. We then
used GATK HaplotypeCaller [@McKenna:2010; @Depristo:2011] for variant (i.e. SNP and indel) calling, yielding the 
results summarized in Table 1.

| Phenotype | Bulk size | Total read bases (bp) | Total reads | GC(%) | Q20(%) | Q30(%) | Coverage a, b | Variants   |
|-----------|-----------|-----------------------|-------------|-------|--------|--------|---------------|------------|
| _EF_      | 11        | 60,303,831,724        | 399,363,124 | 36.93 | 95.02  | 89.25  | 123, 108      | 40,224,519 |
| _IF_      | 8         | 56,804,209,216        | 376,186,816 | 36.92 | 94.84  | 88.94  | 116, 103      | 43,785,856 |
| _LF_      | 11        | 54,587,863,530        | 361,509,030 | 37.14 | 96.00  | 91.34  | 112, 100      | 42,852,937 |
| _NF_      | 9         | 54,296,890,456        | 359,582,056 | 37.13 | 96.84  | 92.81  | 111, 99       | 42,213,427 |

_Table 1. Summary results of the sequencing of bulks of early (_EF_), intermediate (_IF_), late (_LF_) and
"non" flowering (_NF_, actually not flowering at time of DNA extraction) phenotypes. Bulk size refers to the
number of individuals pooled for that phenotype. Coverage is given as a) total read bases divided by reference 
genome size; and b) average mapped coverage._

> #### Supplementary info
> - Input files are deposited at the SRA as project [PRJNA564368](https://www.ncbi.nlm.nih.gov/bioproject/PRJNA564368)
> - The processed reference genome (sans scaffolds; chromosomes only) is available from the main data archive
>   of this study, deposited at Zenodo as [10.5281/zenodo.3402201](http://doi.org/10.5281/zenodo.3402201)
> - Reference mapping was automated with [bwa.sh](../script/bwa.sh). 
> - SNP calling was automated with [snp.sh](../script/snp.sh) 
> - The resulting alignments (sorted BAM files) and variants (compressed VCF files) are  available from Zenodo under 
>   the following locations:
>   - EF [10.5281/zenodo.3401590](http://doi.org/10.5281/zenodo.3401590)
>   - IF [10.5281/zenodo.3401703](http://doi.org/10.5281/zenodo.3401703)
>   - LF [10.5281/zenodo.3401731](http://doi.org/10.5281/zenodo.3401731)
>   - NF [10.5281/zenodo.3402013](http://doi.org/10.5281/zenodo.3402013)
