### DNA extraction, sequencing and mapping

DNA extraction was done by ...
Library preparation was done using TruSeq DNA PCR Free library kit ...
Equimolar concentrations of the F1 genotypes were then pooled into the four bulks _EF_, _IF_, _LF_.
The bulks were sequenced on the Illumina HiSeq X platform resulting in a raw coverage, calculated as 
the total number of read bases divided by the size of the reference genome [approx. 488Mbp; @Parkin:2014], in the 
range of 111x-123x per bulk, i.e. approximately 10x per individual. We used BWA-MEM [@Li:2013] to map the reads 
for each bulk against the reference genome v2.1 of EnsemblPlants release 39, which we had filtered to retain 
chromosomes only. We retained most of the raw coverage in the assemblies, which yielded a mapped coverage in the
range of 99x-108x (Table 1).

| Phenotype | Bulk size | Total read bases (bp) | Total reads | GC(%) | AT(%) | Q20(%) | Q30(%) | Coverage a, b |
|-----------|-----------|-----------------------|-------------|-------|-------|--------|--------|---------------|
| _EF_      | 11        | 60,303,831,724        | 399,363,124 | 36.93 | 63.07 | 95.02  | 89.25  | 123, 108      |
| _IF_      | 8         | 56,804,209,216        | 376,186,816 | 36.92 | 63.08 | 94.84  | 88.94  | 116, 103      |
| _LF_      | 11        | 54,587,863,530        | 361,509,030 | 37.14 | 62.86 | 96.00  | 91.34  | 112, 100      |
| _NF_      | 9         | 54,296,890,456        | 359,582,056 | 37.13 | 62.87 | 96.84  | 92.81  | 111, 99       |

_Table 1. Summary results of the sequencing of bulks of early (_EF_), intermediate (_IF_), late (_LF_) and
"non" flowering (_NF_, actually not flowering at time of DNA extraction) phenotypes. Coverage is given as
a) total read bases divided by reference genome size; and b) average mapped coverage._

> #### Supplementary info
> - The input files are deposited at the SRA under project `PRJNA564368`
> - The processed reference genome (sans scaffolds; chromosomes only) is available here: ...
>
> The mapping was automated using 
> [this](https://github.com/naturalis/brassica-snps/blob/befae9d034f84dd1e124cf80ce639e88b8501baa/script/bwa.sh)
> script. The resulting alignments (sorted BAM files) are available from Zenodo under the following locations:
> - EF
> - IF
> - LF
> - NF
