### Bulk genotyping and QTL region analysis

Given that we phenotyped the F1s by flowering time and sorted into four bulks there are six contrasts among them
(i.e. _EF_ ↔ _IF_, _EF_ ↔ _LF_, _EF_ ↔ _NF_; _IF_ ↔ _LF_, _IF_ ↔ _NF_; and _LF_ ↔ _NF_). We performed joint genotyping
for these contrasts using the GATK CombineGVCFs/GenotypeGVCFs workflow. We then filtered the SNPs further, excluding
low coverage per sample (<40x), low coverage across the pair of merged samples (<100x), high coverage (>400x, e.g. 
repeats), low values for the GATK Genotype Quality score (<99), and low values for the frequency of the reference
allele (<0.2, a conservative value given that we consider TO1000 more or less homozygous). We calculated smoothed 
_G_ statistics [_G'_, see @Magwene:2011] over a sliding window 1Mb wide, filtering outliers by Δ(SNP)-index 
[@Takagi:2013] and retaining all SNPs with _G'_>2.5 for further analysis. We then performed a QTL-seq analysis
[@Takagi:2013] to identify candidate QTL regions by simulation (10k replicates, two-sided 95% confidence interval).

> #### Supplementary info
> **Joint genotyping** - An important issue that needs to be taken care of is that the BAM data (and subsequently 
> the VCF data) that go into the joint genotyping are properly annotated by "read group" (`@RG`), which is the field 
> in SAM/BAM files that is used for identifiers of samples. This so that when `CombineGVCFs` merges the VCF files for 
> the two bulks in such a way that the variants can be traced back to the bulk in which they were observed once
> GenotypeGVCFs does its thing. This is why the extra `-R` argument with a read group ID is passed into BWA-MEM 
> [when doing the assembly](../script/bwa.sh). With that properly taken care of, the GATK `CombineGVCFs`/`GenotypeGVCFs`
> workflow was then automated using [genotype.pl](../script/genotype.pl).

