### Bulk genotyping and QTL region analysis

Given that we phenotyped the F1s by flowering time binned into four bulks there are six contrasts among them
(i.e. _EF_ ↔ _IF_, _EF_ ↔ _LF_, _EF_ ↔ _NF_; _IF_ ↔ _LF_, _IF_ ↔ _NF_; and _LF_ ↔ _NF_). We performed joint genotyping
for these contrasts using the GATK CombineGVCFs/GenotypeGVCFs workflow. We then filtered these genotypes further, 
excluding low coverage per sample (<40x), low coverage across the pair of merged samples (<100x), unusually high 
coverage (>400x, e.g. repeats), low values for the GATK Genotype Quality score (<99), and low values for the frequency 
of the reference allele (<0.2, a conservative value given that we consider TO1000 more or less homozygous). 

We calculated smoothed _G_ statistics [_G'_, see @Magwene:2011] over a sliding window 1Mb wide, filtering outliers by 
Δ(SNP)-index [@Takagi:2013] and retaining all SNPs with _G'_>2.5 for further analysis. We then performed a QTL-seq 
analysis [@Takagi:2013] to identify candidate QTL regions by simulation (10k replicates, two-sided 95% confidence 
interval). For the _G'_ and QTL-seq calculations and simulations we used the _R_ package QTLseqr [@Mansfeld:2018]. 

Based on our inferred QTL regions and smoothed _G'_ values we then scanned the mapped assembly of each bulk for genes 
that fall within QTL regions and have non-synonymous SNPs with high _G'_. Gene coordinates were based on the annotation 
of the TO1000 [i.e. the _Brassica oleracae_ GFF3 release v2.1.39 of EnsemblPlants; @Parkin:2014]. To be able to 
cross-reference the products of these genes with other information resources we then mapped the 
_Brassica oleracea_ genes to curated and machine predicted proteomics identifiers [UniProtKB/TrEMBL; @Schneider:2009]  
using BioMart [@Smedley:2009].

> #### Supplementary info
> The steps described here operate in parallel on all contrasts. The files produced are made available as the archive
> [contrasts.zip](contrasts.zip) in the main data supplement. In it, there are subfolders for the contrasts, named
> `EF-IF`, `EF-LF`, `EF-NF`, `IF-LF`, `IF-NF` and `LF-NF`. The files inside these subfolders (i.e. 
> `${DATA}/contrasts/${CONTRAST}/*`) all follow the same naming scheme, which is described further in the following 
> sections. 
>
> **Joint genotyping** - An important issue that needs to be taken care of is that the BAM data (and subsequently 
> the VCF data) that go into the joint genotyping are properly annotated by "read group" (`@RG`), which is the field 
> in SAM/BAM files that is used for identifiers of samples. This so that when CombineGVCFs merges the VCF files for 
> the two bulks this is done in such a way that the variants can be traced back to the bulk in which they were observed 
> once GenotypeGVCFs does its thing. This is why the extra `-R` argument with a read group ID was passed into BWA-MEM 
> [when doing the assembly](../script/bwa.sh). With that properly taken care of, the GATK CombineGVCFs/GenotypeGVCFs
> workflow was then automated using [genotype.pl](../script/genotype.pl). The files produced by these steps are named:
> - `${DATA}/contrasts/${CONTRAST}/combined-snps.vcf` - Output from VCF merge
> - `${DATA}/contrasts/${CONTRAST}/joint-genotypes.vcf` - Results from joint genotyping 
>
> **QTL analysis** - To get the data into R we needed to transform the GVCF files that GATK produces into tab-separated
> tables, which we did using [qtlseqr.sh](../script/qtlseqr.sh). The actual calculations performed in R are shown in
> the script [QTLseqr.R](../script/QTLseqr.R). The files produced by these steps are named:
> - `${DATA}/contrasts/${CONTRAST}/SNPs_from_GATK.table` - SNPs as TSV
> - `${DATA}/contrasts/${CONTRAST}/SNPs-gprime2.5.csv` - SNPs after filtering
> - `${DATA}/contrasts/${CONTRAST}/gprime.png` - Visualization of QTL regions
> - `${DATA}/contrasts/${CONTRAST}/QTL-regions.csv` - QTL region coordinates
>
> **Database construction** - To be able to reconcile the observed SNPs with gene coordinates, we created a 
> relational database with the following [SQLite schema](../sql/snps.sql), into which we imported data tables for 
> chromosomes (names, centromere locations), genomic features (i.e. locations of genes, 3'/5' UTRs, exons, CDSs, from 
> the TO1000 reference genome annotation), linkage maps (markers, fwd/rev primer sequences, locations in cM) and SNPs 
> (locations, ref/alt allele, numerous summary statistics including G' values). From the database schema we then 
> generated an [object-relational API](../lib) using [DBIx::Class](../sql/make_dbix_api.sh) to provide programmable 
> access to the integrated, indexed data such that downstream analysis scripts have simplified, quicker query access.
> The following files went into populating the database:
> - `${DATA}/reference/sqlite/chromosomes.tsv` - created from GFF3
> - `${DATA}/reference/sqlite/features.tsv` - created from GFF3 using [make_features.pl](../sql/make_features.pl)
> - `${DATA}/reference/sqlite/qtl_regions.tsv` - created from `*/QTL-regions.csv` +
>   [make_qtl_regions.pl](../sql/make_qtl_regions.pl)
> - `${DATA}/reference/sqlite/snps.tsv` - created from `*/SNPs-gprime2.5.csv` + [make_snps.pl](../sql/make_snps.pl)
> - `${DATA}/reference/sqlite/linkages.tsv` - explained [here](../results/linkages) (**Note** - the linkages are not
>   directly relevant to the present paper, but we might need them later for genotyping in other experiments.)
>
> **Finding QTL genes with nonsynonymous substitutions** - For each contrast, we then 
> [queried](../script/genes_in_qtl_regions.pl) which genes intersect with the inferred QTL regions, creating a 
> [list](../results/genes.txt) of 14257 _B. oleracea_ gene IDs sorted by the number of contrasts in which they appear in a 
> QTL region. For each of these genes, we then [calculated](../script/snps_in_cds.pl) whether they contain 
> non-synonymous [SNPs in their coding regions](../results/snps.tsv).
>
> Then, for each contrast, we extracted the gene identifiers that contain nonsynonymous SNPs and translated these to
> UniProtKB/TrEMBL identifiers using [biomart.pl](../script/biomart.pl), resulting in files named as:
> - `${DATA}/contrasts/${CONTRAST}/uniprot.txt` - UniProtKB/TrEMBL IDs
