### Bulk genotyping and QTL region analysis

Given that we phenotyped the F1s by flowering time into four bulks there are six contrasts among them
(i.e. _EF_↔_IF_, _EF_↔_LF_, _EF_↔_NF_; _IF_↔_LF_, _IF_↔_NF_; and _LF_↔_NF_). We performed joint genotyping
for all contrasts using the GATK CombineGVCFs/GenotypeGVCFs workflow.


> #### Supplementary info
> **Joint genotyping** - An important issue that needs to be taken care of is that the BAM data (and subsequently 
> the VCF data) that go into the joint genotyping are properly annotated by "read group" (RG), which is the field 
> in SAM/BAM files that is used for identifiers of samples. This so that when CombineGVCFs merges the VCF files for 
> the two bulks in such a way that the variants can be traced back to the bulk in which they were observed once
> GenotypeGVCFs does its thing. This is why the extra `-R` argument with a read group ID is passed into BWA-MEM 
> when doing the assembly. With that properly taken care of, the GATK CombineGVCFs/GenotypeGVCFs workflow was then
> automated using [genotype.pl](../script/genotype.pl).
